Code/network/src/Connection.cpp

```diff
@@ -126,28 +222,77 @@ void Connection::Update(uint64_t aElapsedMilliseconds)
     default:
         break;
     }
+
+    return m_state;
 }
 
-void Connection::SendNegotiation()
+void Connection::WriteHeader(Buffer::Writer& aWriter, const uint64_t acHeaderType)
 {
     Header header;
     header.Signature[0] = s_headerSignature[0];
     header.Signature[1] = s_headerSignature[1];
     header.Version = 1;
-    header.Type = Header::kNegotiation;
+    header.Type = acHeaderType;
     header.Length = 0;
 
+    aWriter.WriteBytes((const uint8_t*)header.Signature, 2);
+    aWriter.WriteBits(header.Version, 6);
+    aWriter.WriteBits(header.Type, 3);
+    aWriter.WriteBits(header.Length, 11);
+}
+
+void Connection::Disconnect()
+{
+    StackAllocator<256> allocator;
+    auto* pBuffer = allocator.New<Buffer>(16);
+
+    Buffer::Writer writer(pBuffer);
+    WriteHeader(writer, Header::kDisconnect);
+
+    // we don't take remote code into account here as the negotiation may now have been successful yet
+    uint32_t codeToSend = m_challengeCode;
+    m_filter.PostSend((uint8_t *)&codeToSend, sizeof(codeToSend), UINT32_MAX);
+    WriteChallenge(writer, codeToSend);
+
+    for (uint8_t i = 0; i < 10; i++)
+    {
+        // send a bunch of them so they have more chances of reaching the remote
+        m_communication.Send(m_remoteEndpoint, *pBuffer);
+    }
+
+    m_state = kNone;
+    allocator.Delete(pBuffer);
+}
+
+void Connection::SendNegotiation()
+{
     StackAllocator<1 << 13> allocator;
-    auto* pBuffer = allocator.New<Buffer>(1200);
+    auto* pBuffer = allocator.New<Buffer>(m_isServer ? 200 : 1200);

```

Comment:
It's allocated on the stack, don't need to optimize the size

Reply:
This is not about optimization, but to prevent the server being useful for reflection in a DDOS attack,. The gafferon article I've linked also speaks about amplification, but if the packets sizes are the same there's no real amplification going on.

Reply:
BTW it's not actually stack allocated, as `Buffer` still uses its default `StandardAllocator` internally for `m_pData`. Want me to fix that? I could give it an optional `apAllocator` parameter in the constructor.